File: materias.php
<?php
require 'menu.php';
require 'db.php';

require_once __DIR__ . '/lib/security_headers.php';

$uid = (int)$_SESSION['usuarioId'];
$userRow = dbOne("SELECT nombre, rol, ultimaMateria FROM usuarios WHERE usuarioId = :id", ['id' => $uid]);
$first   = htmlspecialchars(explode(' ', $userRow['nombre'] ?? 'Usuario')[0]);
$rol     = $_navRol ?? ($userRow['rol'] ?? 'estudiante');
$isTeacher = ($rol !== 'estudiante' && $rol !== 'student');
$ultimaMateria = (int)($userRow['ultimaMateria'] ?? 0);

$colors = [
    1=>'#2563EB',2=>'#059669',3=>'#7C3AED',4=>'#0284C7',5=>'#D97706',
    6=>'#0D9488',7=>'#DC2626',8=>'#DB2777',9=>'#EA580C',10=>'#0891B2',11=>'#E11D48',
];
$icons = [
    1=>'hash',2=>'activity',3=>'zap',4=>'cpu',5=>'book-open',
    6=>'map',7=>'feather',8=>'globe',9=>'pen-tool',10=>'monitor',11=>'heart',
];

$subjects = dbAll(
    "SELECT m.materiaId AS id, m.nombre,
            (SELECT COUNT(*) FROM clases_programadas cp WHERE cp.materiaId = m.materiaId AND cp.activa = true) AS clases_activas
     FROM materias m ORDER BY m.nombre"
);
if (empty($subjects)) {
    $names = ['Mathematics','Biology','Chemistry','Physics','History','Geography','Literature','Foreign Languages','Art and Music','Technology','Physical Education'];
    foreach ($names as $i => $n) $subjects[] = ['id' => $i + 1, 'nombre' => $n, 'clases_activas' => 0];
}
?>
<div class="ml-wrap">
  <div class="ml-wrap-inner">
  <div class="ml-head" style="align-items:flex-end">
    <div>
      <div style="font-size:13px;color:var(--sub);font-weight:400;margin-bottom:2px">¡Hola, <?= $first ?>!</div>
      <div class="ml-head-title" style="margin:0">¿Qué estudias hoy?</div>
    </div>
    <?php if ($isTeacher): ?>
    <button class="ml-btn" style="width:44px;height:44px;padding:0;justify-content:center;border-radius:14px;background:var(--pb);color:var(--p)" onclick="location.href='dashboard_profesor.php'">
      <i data-feather="bar-chart-2" style="width:20px;height:20px"></i>
    </button>
    <?php endif; ?>
  </div>

  <?php
  $continuar = null;
  if ($ultimaMateria >= 1 && $ultimaMateria <= 11) {
      foreach ($subjects as $s) {
          if ((int)$s['id'] === $ultimaMateria) {
              $continuar = $s;
              break;
          }
      }
  }
  ?>

  <div class="ml-grid">
    <?php if ($continuar): $cc = $colors[$ultimaMateria]; $ico = $icons[$ultimaMateria]; ?>
    <a href="contenido.php?materia=<?= $ultimaMateria ?>&nombre=<?= urlencode($continuar['nombre']) ?>" style="text-decoration:none;display:flex;flex-direction:column;align-items:center;justify-content:center;border-radius:18px;padding:20px 16px;background:<?= $cc ?>;min-height:150px;color:#fff;box-shadow:0 4px 14px rgba(0,0,0,.12);border:2px solid #fff;position:relative">
      <div style="position:absolute;top:12px;left:12px;padding:4px 12px;border-radius:20px;background:rgba(0,0,0,.25);color:#fff;font-size:11px;font-weight:600;letter-spacing:.5px">CONTINUAR</div>
      <div style="width:64px;height:64px;border-radius:18px;background:rgba(255,255,255,.22);display:flex;align-items:center;justify-content:center;margin-bottom:12px">
        <i data-feather="corner-left-up" style="width:30px;height:30px;color:#fff"></i>
      </div>
      <div style="font-weight:600;font-size:14px;line-height:1.3;text-align:center;color:#fff"><?= htmlspecialchars($continuar['nombre']) ?></div>
    </a>
    <?php endif; ?>
    <?php foreach ($subjects as $s):
      $id = (int)$s['id'];
      $c = $colors[$id] ?? '#888';
      $ico = $icons[$id] ?? 'book';
      $active = (int)($s['clases_activas'] ?? 0);
    ?>
    <a href="contenido.php?materia=<?= $id ?>&nombre=<?= urlencode($s['nombre']) ?>" style="text-decoration:none;display:flex;flex-direction:column;align-items:center;border-radius:18px;padding:20px 16px;background:<?= $c ?>;min-height:150px;color:#fff;box-shadow:0 4px 14px rgba(0,0,0,.12)">
      <div style="width:64px;height:64px;border-radius:18px;background:rgba(255,255,255,.22);display:flex;align-items:center;justify-content:center;margin-bottom:12px">
        <i data-feather="<?= $ico ?>" style="width:34px;height:34px;color:#fff"></i>
      </div>
      <div style="font-weight:600;font-size:14px;line-height:1.3;text-align:center;color:#fff"><?= htmlspecialchars($s['nombre']) ?></div>
      <?php if ($active > 0): ?>
      <div style="margin-top:8px;padding:3px 10px;border-radius:20px;background:rgba(255,255,255,.25);color:#fff;font-size:11px;font-weight:500;align-self:center"><?= $active ?> en vivo</div>
      <?php endif; ?>
    </a>
    <?php endforeach; ?>
  </div>
  </div>
</div>
<?php require 'footer.php'; ?>

File: go/internal/web/materias.go
package web

import (
	"fmt"
	"net/http"
	"net/url"
	"strings"

	"classexpress/internal/i18n"
	"classexpress/internal/store"
)

var subjectColors = map[int64]string{
	1: "#2563EB", 2: "#059669", 3: "#7C3AED", 4: "#0284C7", 5: "#D97706",
	6: "#0D9488", 7: "#DC2626", 8: "#DB2777", 9: "#EA580C", 10: "#0891B2", 11: "#E11D48",
}

var subjectIcons = map[int64]string{
	1: "hash", 2: "activity", 3: "zap", 4: "cpu", 5: "book-open",
	6: "map", 7: "feather", 8: "globe", 9: "pen-tool", 10: "monitor", 11: "heart",
}

type subjectItem struct {
	ID       int64
	Nombre   string
	NombreURL string
	Color    string
	Icon     string
	Activas  int64
}

// HandleMaterias ports materias.php.
// Supports both logged-in users and guests (via remember_token or role param).
func (p *Pages) HandleMaterias(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	s := SessionFrom(ctx)
	lang := p.ResolveLang(nil, r) // Allow lang without session

	// Check for role parameter (from landing page redirect)
	roleParam := r.URL.Query().Get("rol")
	if roleParam != "" && (roleParam == "student" || roleParam == "instructor") {
		http.SetCookie(w, &http.Cookie{
			Name:     "ce_role",
			Value:    roleParam,
			Path:     "/",
			MaxAge:   365 * 24 * 60 * 60,
			SameSite: http.SameSiteLaxMode,
		})
	}

	// Try to get session (may be nil for new guests)
	if s == nil {
		// Try to auto-login via remember_token
		if cookie, err := r.Cookie("ce_remember"); err == nil && cookie.Value != "" {
			ctx := r.Context()
			row, err := p.DB.QueryOne(ctx,
				"SELECT usuarioId, nombre, rol, creditos, idioma_preferido FROM usuarios WHERE remember_token = ? AND remember_token IS NOT NULL AND eliminado = 0 LIMIT 1",
				HashToken(cookie.Value))
			if err == nil && row != nil {
				s = &Session{ID: newSessionID(), Values: map[string]string{}, IsNew: true}
				s.Set("usuarioId", store.Str(row["usuarioId"]))
				s.Set("nombre", store.Str(row["nombre"]))
				s.Set("rol", store.Str(row["rol"]))
				s.Set("creditos", store.Str(row["creditos"]))
				if lang := store.Str(row["idioma_preferido"]); lang != "" {
					s.Set("_lang", lang)
				}
				newTok := NewRememberToken()
				_, _ = p.DB.Exec(ctx, "UPDATE usuarios SET remember_token = ? WHERE usuarioId = ?", HashToken(newTok), store.Str(row["usuarioId"]))
				http.SetCookie(w, &http.Cookie{Name: "ce_remember", Value: newTok, Path: "/", MaxAge: 30 * 24 * 60 * 60, HttpOnly: true, Secure: IsHTTPS(r), SameSite: http.SameSiteLaxMode})
				http.SetCookie(w, &http.Cookie{Name: CookieName, Value: s.ID, Path: "/", MaxAge: 30 * 24 * 60 * 60, HttpOnly: true, Secure: IsHTTPS(r), SameSite: http.SameSiteLaxMode})
			}
		}
	}

	// Get nav data (works with nil session)
	nav, redirect := p.MenuData(w, r, nil, "materias.php", lang)
	if redirect {
		return
	}
	nav.AAAdUnitID = p.Cfg.AAAdUnitID

	// Determine user state
	first := "Usuario"
	ultimaMateria := int64(0)

	if s != nil && LoggedIn(s) {
		userRow, err := p.DB.QueryOne(r.Context(), "SELECT nombre, rol, ultimaMateria FROM usuarios WHERE usuarioId = ?", UID(s))
		if err == nil && userRow != nil {
			n := store.Str(userRow["nombre"])
			if n != "" {
				first = n
				if i := strings.IndexByte(first, ' '); i > 0 {
					first = first[:i]
				}
			}
			if store.Str(userRow["rol"]) == "instructor" {
				nav.IsTeacher = true
			}
			ultimaMateria = store.Int(userRow["ultimaMateria"])
		}
	} else {
		if roleCookie, _ := r.Cookie("ce_role"); roleCookie != nil && roleCookie.Value == "instructor" {
			nav.IsTeacher = true
		}
	}

	// Load subjects
	subjects, err := p.DB.QueryAll(r.Context(),
		`SELECT m.materiaId AS id, m.nombre,
		        (SELECT COUNT(*) FROM clases_programadas cp WHERE cp.materiaId = m.materiaId AND cp.activa = true) AS clases_activas
		 FROM materias m ORDER BY m.nombre`)
	if err != nil || len(subjects) == 0 {
		names := []string{"Mathematics", "Biology", "Chemistry", "Physics", "History", "Geography", "Literature", "Foreign Languages", "Art and Music", "Technology", "Physical Education"}
		subjects = []map[string]any{}
		for i, n := range names {
			subjects = append(subjects, map[string]any{"id": i + 1, "nombre": n, "clases_activas": 0})
		}
	}

	var items []subjectItem
	for _, s := range subjects {
		id := store.Int(s["id"])
		nombre := store.Str(s["nombre"])
		if t := i18n.T(lang, "subject.name."+fmt.Sprint(id), nil); t != "" {
			nombre = t
		}
		items = append(items, subjectItem{
			ID:        id,
			Nombre:    nombre,
			NombreURL: url.QueryEscape(nombre),
			Color:     subjectColors[id],
			Icon:      subjectIcons[id],
			Activas:   store.Int(s["clases_activas"]),
		})
	}

	// Continuar card
	if s != nil && LoggedIn(s) {
		if row, _ := p.DB.QueryOne(r.Context(), "SELECT nombre FROM usuarios WHERE usuarioId = ?", UID(s)); row != nil {
			n := store.Str(row["nombre"])
			if n != "" {
				first = n
				if i := strings.IndexByte(first, ' '); i > 0 {
					first = first[:i]
				}
			}
		}
	}

isTeacher := nav.IsTeacher

data := map[string]any{
		"Lang":          lang,
		"NavData":       nav,
		"First":         first,
		"IsTeacher":     isTeacher,
		"Subjects":      items,
		"UltimaMateria": ultimaMateria,
	}

	if err := p.Templates.RenderAuthed(w, "materias", p, s, lang, data); err != nil {
		serverError(w, err)
	}
}

func indexByte(s string, b byte) int {
	for i := 0; i < len(s); i++ {
		if s[i] == b {
			return i
		}
	}
	return -1
}